File: Autonomous_Mobile_Robots_Course/Chapter14/Lesson4/Chapter14_Lesson4.py
"""
Chapter14_Lesson4.py
Recovery Behaviors and Fault Handling — ROS 2 / Nav2-oriented supervisor (example)

Notes:
- This script is a *teaching* example. Topic/service names can differ by robot/nav2 config.
- It demonstrates: progress monitoring, oscillation detection, costmap clearing, simple spin/backup,
  and escalation to safe-stop.

Dependencies (typical):
  sudo apt install ros-<distro>-nav2-msgs ros-<distro>-geometry-msgs ros-<distro>-nav-msgs
  pip install --user numpy
"""

import time
from dataclasses import dataclass

import numpy as np
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy

from geometry_msgs.msg import Twist, PoseWithCovarianceStamped
from nav_msgs.msg import Odometry

# Nav2 message/service types (names may vary by distro)
from nav2_msgs.srv import ClearEntireCostmap


@dataclass
class MonitorParams:
    # Progress monitor
    window_s: float = 10.0          # seconds
    min_progress_m: float = 0.25    # required decrease in distance-to-goal over window
    # Oscillation monitor
    osc_window_s: float = 6.0
    osc_sign_flips: int = 6         # number of sign flips in cmd_vel over window
    v_eps: float = 0.03             # treat |v| < v_eps as zero
    # Localization uncertainty (AMCL covariance trace threshold)
    cov_trace_max: float = 0.35
    # Recovery actions
    spin_time_s: float = 3.0
    spin_wz: float = 0.8
    backup_time_s: float = 2.0
    backup_vx: float = -0.12
    # Escalation
    max_recovery_attempts: int = 3
    # Topics/services (common defaults)
    cmd_vel_topic: str = "/cmd_vel"
    odom_topic: str = "/odom"
    amcl_topic: str = "/amcl_pose"
    local_clear_srv: str = "/local_costmap/clear_entirely_local_costmap"
    global_clear_srv: str = "/global_costmap/clear_entirely_global_costmap"


class RecoverySupervisor(Node):
    def __init__(self):
        super().__init__("recovery_supervisor")

        # Parameters (could be declared as ROS params; kept simple for teaching)
        self.p = MonitorParams()

        qos = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=10
        )

        self.pub_cmd = self.create_publisher(Twist, self.p.cmd_vel_topic, 10)
        self.sub_odom = self.create_subscription(Odometry, self.p.odom_topic, self.on_odom, qos)
        self.sub_amcl = self.create_subscription(PoseWithCovarianceStamped, self.p.amcl_topic, self.on_amcl, 10)
        self.sub_cmd = self.create_subscription(Twist, self.p.cmd_vel_topic, self.on_cmd, 10)

        self.cli_local = self.create_client(ClearEntireCostmap, self.p.local_clear_srv)
        self.cli_global = self.create_client(ClearEntireCostmap, self.p.global_clear_srv)

        # State buffers
        self.goal_xy = None  # (gx, gy) set externally (e.g., from NavigateToPose goal); for demo we keep None
        self.last_xy = None
        self.last_cov_trace = None

        self.progress_hist = []  # list of (t, dist_to_goal)
        self.cmd_hist = []       # list of (t, vx, wz)
        self.recovery_attempts = 0
        self.in_recovery = False

        # Periodic check loop
        self.timer = self.create_timer(0.2, self.tick)

        self.get_logger().info("RecoverySupervisor started. Set goal_xy in code or adapt to listen to Nav2 goals.")

    # ------------------------- Callbacks -------------------------

    def on_odom(self, msg: Odometry):
        x = msg.pose.pose.position.x
        y = msg.pose.pose.position.y
        self.last_xy = (x, y)

        if self.goal_xy is not None:
            d = float(np.hypot(self.goal_xy[0] - x, self.goal_xy[1] - y))
            t = self.now_s()
            self.progress_hist.append((t, d))
            self.progress_hist = self.trim_by_time(self.progress_hist, self.p.window_s)

    def on_amcl(self, msg: PoseWithCovarianceStamped):
        cov = np.array(msg.pose.covariance, dtype=float).reshape(6, 6)
        self.last_cov_trace = float(np.trace(cov[:3, :3]))  # x,y,yaw block approx

    def on_cmd(self, msg: Twist):
        t = self.now_s()
        self.cmd_hist.append((t, float(msg.linear.x), float(msg.angular.z)))
        self.cmd_hist = self.trim_by_time(self.cmd_hist, self.p.osc_window_s)

    # ------------------------- Helpers -------------------------

    def now_s(self) -> float:
        return self.get_clock().now().nanoseconds * 1e-9

    @staticmethod
    def trim_by_time(hist, window_s):
        if not hist:
            return hist
        t_latest = hist[-1][0]
        t_min = t_latest - window_s
        i = 0
        while i < len(hist) and hist[i][0] < t_min:
            i += 1
        return hist[i:]

    def progress_ok(self) -> bool:
        """Return True if sufficient progress toward goal occurred in the window."""
        if self.goal_xy is None or len(self.progress_hist) < 2:
            return True  # not enough info; don't fault
        d_old = self.progress_hist[0][1]
        d_new = self.progress_hist[-1][1]
        return (d_old - d_new) >= self.p.min_progress_m

    def oscillating(self) -> bool:
        """Detect oscillation by counting sign flips of linear-x or angular-z commands."""
        if len(self.cmd_hist) < 3:
            return False

        def sign_eps(v):
            if abs(v) < self.p.v_eps:
                return 0
            return 1 if v > 0 else -1

        sx = [sign_eps(vx) for _, vx, _ in self.cmd_hist]
        sz = [sign_eps(wz) for _, _, wz in self.cmd_hist]

        flips_x = sum(1 for i in range(1, len(sx)) if sx[i] != 0 and sx[i-1] != 0 and sx[i] != sx[i-1])
        flips_z = sum(1 for i in range(1, len(sz)) if sz[i] != 0 and sz[i-1] != 0 and sz[i] != sz[i-1])

        return (flips_x + flips_z) >= self.p.osc_sign_flips

    def localization_bad(self) -> bool:
        if self.last_cov_trace is None:
            return False
        return self.last_cov_trace > self.p.cov_trace_max

    # ------------------------- Recovery actions -------------------------

    def call_clear_costmaps(self):
        """Best-effort: clear local then global costmap."""
        for cli, name in [(self.cli_local, "local"), (self.cli_global, "global")]:
            if not cli.service_is_ready():
                self.get_logger().warn(f"Clear {name} costmap service not ready: {cli.srv_name}")
                continue
            req = ClearEntireCostmap.Request()
            fut = cli.call_async(req)
            rclpy.spin_until_future_complete(self, fut, timeout_sec=2.0)
            self.get_logger().info(f"Clear {name} costmap result: {fut.result() is not None}")

    def publish_for(self, vx: float, wz: float, duration_s: float):
        """Open-loop velocity command for a limited time."""
        t_end = time.time() + duration_s
        msg = Twist()
        while rclpy.ok() and time.time() < t_end:
            msg.linear.x = float(vx)
            msg.angular.z = float(wz)
            self.pub_cmd.publish(msg)
            time.sleep(0.05)
        # stop
        msg.linear.x = 0.0
        msg.angular.z = 0.0
        self.pub_cmd.publish(msg)

    def do_recovery_cycle(self, reason: str):
        self.in_recovery = True
        self.recovery_attempts += 1

        self.get_logger().warn(f"RECOVERY #{self.recovery_attempts}: reason={reason}")

        # 1) Clear costmaps (handles transient obstacles / stale inflation)
        self.call_clear_costmaps()

        # 2) Spin in place (improves sensor coverage / localization)
        self.publish_for(0.0, self.p.spin_wz, self.p.spin_time_s)

        # 3) Backup (escape tight corners / local minima near obstacles)
        self.publish_for(self.p.backup_vx, 0.0, self.p.backup_time_s)

        self.in_recovery = False

    def safe_stop(self, reason: str):
        self.get_logger().error(f"SAFE STOP: {reason}")
        msg = Twist()
        msg.linear.x = 0.0
        msg.angular.z = 0.0
        # publish a few times
        for _ in range(10):
            self.pub_cmd.publish(msg)
            time.sleep(0.05)

    # ------------------------- Main tick -------------------------

    def tick(self):
        if self.in_recovery:
            return

        if self.recovery_attempts >= self.p.max_recovery_attempts:
            self.safe_stop("exceeded max recovery attempts")
            rclpy.shutdown()
            return

        # Fault conditions (simple; extend with more residuals)
        if self.localization_bad():
            self.do_recovery_cycle("localization covariance too high")
            return

        if self.goal_xy is not None:
            if (not self.progress_ok()) or self.oscillating():
                why = "no progress" if not self.progress_ok() else "oscillation"
                self.do_recovery_cycle(why)
                return


def main():
    rclpy.init()
    node = RecoverySupervisor()

    # DEMO: set a fixed goal (replace with nav2 action goal listener)
    node.goal_xy = (5.0, 0.0)

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: Autonomous Mobile Robots Course/Chapter14/Lesson4/Chapter14_Lesson4.cpp
/* 
Chapter14_Lesson4.cpp
Recovery Behaviors and Fault Handling — ROS 2 / Nav2-oriented supervisor (example)

Build example (adapt your ROS 2 distro):
  add_executable(recovery_supervisor src/Chapter14_Lesson4.cpp)
  ament_target_dependencies(recovery_supervisor rclcpp geometry_msgs nav_msgs nav2_msgs)
*/

#include <chrono>
#include <cmath>
#include <deque>
#include <string>
#include <utility>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
#include "nav2_msgs/srv/clear_entire_costmap.hpp"

using namespace std::chrono_literals;

struct MonitorParams {
  double window_s = 10.0;
  double min_progress_m = 0.25;

  double osc_window_s = 6.0;
  int osc_sign_flips = 6;
  double v_eps = 0.03;

  double cov_trace_max = 0.35;

  double spin_time_s = 3.0;
  double spin_wz = 0.8;
  double backup_time_s = 2.0;
  double backup_vx = -0.12;

  int max_recovery_attempts = 3;

  std::string cmd_vel_topic = "/cmd_vel";
  std::string odom_topic = "/odom";
  std::string amcl_topic = "/amcl_pose";
  std::string local_clear_srv = "/local_costmap/clear_entirely_local_costmap";
  std::string global_clear_srv = "/global_costmap/clear_entirely_global_costmap";
};

class RecoverySupervisor : public rclcpp::Node {
public:
  RecoverySupervisor() : Node("recovery_supervisor_cpp") {
    p_ = MonitorParams{};

    pub_cmd_ = this->create_publisher<geometry_msgs::msg::Twist>(p_.cmd_vel_topic, 10);

    sub_odom_ = this->create_subscription<nav_msgs::msg::Odometry>(
      p_.odom_topic, rclcpp::QoS(10).best_effort(),
      std::bind(&RecoverySupervisor::on_odom, this, std::placeholders::_1));

    sub_amcl_ = this->create_subscription<geometry_msgs::msg::PoseWithCovarianceStamped>(
      p_.amcl_topic, 10,
      std::bind(&RecoverySupervisor::on_amcl, this, std::placeholders::_1));

    sub_cmd_ = this->create_subscription<geometry_msgs::msg::Twist>(
      p_.cmd_vel_topic, 10,
      std::bind(&RecoverySupervisor::on_cmd, this, std::placeholders::_1));

    cli_local_ = this->create_client<nav2_msgs::srv::ClearEntireCostmap>(p_.local_clear_srv);
    cli_global_ = this->create_client<nav2_msgs::srv::ClearEntireCostmap>(p_.global_clear_srv);

    goal_set_ = true;
    goal_x_ = 5.0;
    goal_y_ = 0.0;

    timer_ = this->create_wall_timer(200ms, std::bind(&RecoverySupervisor::tick, this));
    RCLCPP_INFO(this->get_logger(), "RecoverySupervisor C++ started (demo goal set to (5,0)).");
  }

private:
  // history entries
  using DistEntry = std::pair<double, double>; // (t, dist)
  struct CmdEntry { double t; double vx; double wz; };

  void on_odom(const nav_msgs::msg::Odometry::SharedPtr msg) {
    last_x_ = msg->pose.pose.position.x;
    last_y_ = msg->pose.pose.position.y;
    have_pose_ = true;

    if (goal_set_) {
      double d = std::hypot(goal_x_ - last_x_, goal_y_ - last_y_);
      double t = now_s();
      progress_hist_.push_back({t, d});
      trim_progress_();
    }
  }

  void on_amcl(const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr msg) {
    // covariance is 6x6 row-major in array[36]
    const auto & c = msg->pose.covariance;
    // trace of (x,y,yaw) approx indices: (0,0), (1,1), (5,5)
    last_cov_trace_ = c[0] + c[7] + c[35];
    have_cov_ = true;
  }

  void on_cmd(const geometry_msgs::msg::Twist::SharedPtr msg) {
    double t = now_s();
    cmd_hist_.push_back({t, msg->linear.x, msg->angular.z});
    trim_cmd_();
  }

  double now_s() const {
    return this->get_clock()->now().nanoseconds() * 1e-9;
  }

  void trim_progress_() {
    if (progress_hist_.empty()) return;
    double t_latest = progress_hist_.back().first;
    double t_min = t_latest - p_.window_s;
    while (!progress_hist_.empty() && progress_hist_.front().first < t_min) {
      progress_hist_.pop_front();
    }
  }

  void trim_cmd_() {
    if (cmd_hist_.empty()) return;
    double t_latest = cmd_hist_.back().t;
    double t_min = t_latest - p_.osc_window_s;
    while (!cmd_hist_.empty() && cmd_hist_.front().t < t_min) {
      cmd_hist_.pop_front();
    }
  }

  bool progress_ok_() const {
    if (!goal_set_ || progress_hist_.size() < 2) return true;
    double d0 = progress_hist_.front().second;
    double d1 = progress_hist_.back().second;
    return (d0 - d1) >= p_.min_progress_m;
  }

  int sgn_eps_(double v) const {
    if (std::fabs(v) < p_.v_eps) return 0;
    return (v > 0.0) ? 1 : -1;
  }

  bool oscillating_() const {
    if (cmd_hist_.size() < 3) return false;
    int flips = 0;
    for (size_t i = 1; i < cmd_hist_.size(); ++i) {
      int s_prev = sgn_eps_(cmd_hist_[i-1].vx);
      int s_now  = sgn_eps_(cmd_hist_[i].vx);
      if (s_prev != 0 && s_now != 0 && s_prev != s_now) flips++;
    }
    for (size_t i = 1; i < cmd_hist_.size(); ++i) {
      int s_prev = sgn_eps_(cmd_hist_[i-1].wz);
      int s_now  = sgn_eps_(cmd_hist_[i].wz);
      if (s_prev != 0 && s_now != 0 && s_prev != s_now) flips++;
    }
    return flips >= p_.osc_sign_flips;
  }

  bool localization_bad_() const {
    if (!have_cov_) return false;
    return last_cov_trace_ > p_.cov_trace_max;
  }

  void publish_for_(double vx, double wz, double duration_s) {
    auto t_end = this->get_clock()->now() + rclcpp::Duration::from_seconds(duration_s);
    rclcpp::Rate rate(20.0);
    geometry_msgs::msg::Twist cmd;
    while (rclcpp::ok() && this->get_clock()->now() < t_end) {
      cmd.linear.x = vx;
      cmd.angular.z = wz;
      pub_cmd_->publish(cmd);
      rate.sleep();
    }
    cmd.linear.x = 0.0; cmd.angular.z = 0.0;
    pub_cmd_->publish(cmd);
  }

  void call_clear_costmaps_() {
    for (auto & cli : {cli_local_, cli_global_}) {
      if (!cli->service_is_ready()) {
        RCLCPP_WARN(this->get_logger(), "Costmap clear service not ready: %s", cli->get_service_name());
        continue;
      }
      auto req = std::make_shared<nav2_msgs::srv::ClearEntireCostmap::Request>();
      auto fut = cli->async_send_request(req);
      auto ret = rclcpp::spin_until_future_complete(this->get_node_base_interface(), fut, 2s);
      RCLCPP_INFO(this->get_logger(), "Clear costmap %s", (ret == rclcpp::FutureReturnCode::SUCCESS) ? "OK" : "TIMEOUT/FAIL");
    }
  }

  void do_recovery_(const std::string & reason) {
    in_recovery_ = true;
    recovery_attempts_++;

    RCLCPP_WARN(this->get_logger(), "RECOVERY #%d reason=%s", recovery_attempts_, reason.c_str());

    call_clear_costmaps_();
    publish_for_(0.0, p_.spin_wz, p_.spin_time_s);
    publish_for_(p_.backup_vx, 0.0, p_.backup_time_s);

    in_recovery_ = false;
  }

  void safe_stop_(const std::string & reason) {
    RCLCPP_ERROR(this->get_logger(), "SAFE STOP: %s", reason.c_str());
    geometry_msgs::msg::Twist cmd;
    cmd.linear.x = 0.0;
    cmd.angular.z = 0.0;
    for (int i = 0; i < 10; ++i) {
      pub_cmd_->publish(cmd);
      rclcpp::sleep_for(50ms);
    }
  }

  void tick() {
    if (in_recovery_) return;

    if (recovery_attempts_ >= p_.max_recovery_attempts) {
      safe_stop_("exceeded max recovery attempts");
      rclcpp::shutdown();
      return;
    }

    if (localization_bad_()) {
      do_recovery_("localization covariance too high");
      return;
    }

    if (goal_set_) {
      if (!progress_ok_() || oscillating_()) {
        do_recovery_(!progress_ok_() ? "no progress" : "oscillation");
        return;
      }
    }
  }

private:
  MonitorParams p_;

  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr pub_cmd_;
  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_odom_;
  rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr sub_amcl_;
  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr sub_cmd_;

  rclcpp::Client<nav2_msgs::srv::ClearEntireCostmap>::SharedPtr cli_local_;
  rclcpp::Client<nav2_msgs::srv::ClearEntireCostmap>::SharedPtr cli_global_;

  rclcpp::TimerBase::SharedPtr timer_;

  bool have_pose_ = false;
  double last_x_ = 0.0, last_y_ = 0.0;

  bool have_cov_ = false;
  double last_cov_trace_ = 0.0;

  bool goal_set_ = false;
  double goal_x_ = 0.0, goal_y_ = 0.0;

  std::deque<DistEntry> progress_hist_;
  std::deque<CmdEntry> cmd_hist_;

  int recovery_attempts_ = 0;
  bool in_recovery_ = false;
};

int main(int argc, char ** argv) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<RecoverySupervisor>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
